### PR TITLE
misc: printk: _vprintk: extending parsing functionality.

### LINKED ISSUE #10195: Shell dereferences invalid pointer when printing demo command help
On a shell over uart, typing "demo\r" without any option causes a bad pointer de-refence:

==30417==    at 0x804AD0C: _vprintk (printk.c:196)
==30417==    by 0x804FFA3: shell_fprintf_fmt (shell_fprintf.c:38)
==30417==    by 0x804F7BD: shell_fprintf (shell.c:1409)
==30417==    by 0x804FADC: help_item_print (shell.c:1537)
==30417==    by 0x804FB62: help_options_print (shell.c:1561)
==30417==    by 0x804FEB8: shell_help_print (shell.c:1662)
==30417==    by 0x804EECB: shell_execute (shell.c:1080)
==30417==    by 0x804E87C: shell_state_collect (shell.c:770)
==30417==    by 0x804F69F: shell_process (shell.c:1366)
==30417==    by 0x804F3B5: shell_thread (shell.c:1232)
==30417==  Address 0xa is not stack'd, malloc'd or (recently) free'd

### LINKED ISSUE #10287: Extend _vprintk functionality to parse *
Current `_vprintk` implementation does not support `*`. It is not possible to parse following data:
`"%s%-*s%s:"`.

Fix can be based on `_prf` function localized in /lib/libc/minimal/source/stdout/prf.c

### PR DESCRIPTION
**Pull request by jakub-uC**
1. Function _vprintk is able to parse '*'.
2. Fixed checkpatch warnings.

Fixes #10195 
Fixes #10287

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

### CODE REVIEW

**Comment by carlescufi** [comment]
@jarz-nordic can you please run a quick `ninja rom_report` to know what the impact is in terms of ROM size of this change? This is going to affect nearly every build of Zephyr.

**Review by aescolar — COMMENTED** [review]
I haven't looked carefully in the code or tried it. But what I can see is ~that we don't have any printk specific test, and~ that (logically) the new code is not covered in any test. ~Could you add a testcase for printk covering the different supported formats?~ Could you extend the ```tests/kernel/common/``` printk test with the new formats?

**Comment by jarz-nordic** [comment]
@carlescufi : Parsing `*` adds extra 182 of ROM. I have added `ifdef` to compile it only when SHELL is enabled.

@aescolar : I have added test case to check `*` parsing

**Review by carlescufi — APPROVED** [review]

**Comment by aescolar** [comment]
@jarz-nordic : From the [coverage report ](https://codecov.io/gh/zephyrproject-rtos/zephyr/pull/10299/diff?src=pr&el=tree#diff-bWlzYy9wcmludGsuYw==) you can see a few paths are not covered (red: line not covered ; yellow: branch not covered in both conditions). Would it be possible to expand the test to cover them?

**Review by andyross — CHANGES_REQUESTED** [review]
Sorry... why?  It's already a semi-disaster that we have two different printf() implementations.

Why isn't the solution here to use a sprintf() call into libc instead of extending printk() (which was never intended to be a conformant printf!) yet again?

(Note that things are so bad that when doing x86_64 I found I had to write a *third* printf-like-thing so I could have a header-only one I could use in 16 and 32 bit stub code).

printk is too big already, IMHO.

**Comment by andyross** [comment]
To be less contentious: I don't see anything wrong with the code.  But if the alternatively solution is, as I suspect, to sprintf() the code in all cases instead of using "printk when POSIX, otherwise libc", then that seems 100% preferable to me: much smaller, more robust to other divergences, less overloading printk() to be something other than a simple kernel logging tool.

**Comment by jarz-nordic** [comment]
@andyross : this change has no direct impact on existing code because by default this extra code will not be compiled.
This change is needed in specific case when shell is used with POSIX, otherwise shell uses `_prf` function from `/lib/libc/minimal/source/stdout/prf.c`.

I do not insist to implement this change. If you have any other suggestion how I could fix #10195 in other way please let me know.

**Comment by aescolar** [comment]
> This change is needed in specific case when shell is used with POSIX

You mean: "when compiled without minimal libc"?

**Comment by jarz-nordic** [comment]
Sorry,
To be more precise please take a look at below implementation:

```
#if !defined(CONFIG_NEWLIB_LIBC) && !defined(CONFIG_ARCH_POSIX)
	(void)_prf(out_func, (void *)sh_fprintf, (char *)fmt, args);
#else
	_vprintk(out_func, (void *)sh_fprintf, fmt, args);
#endif
```

**Comment by jarz-nordic** [comment]
Yes, this was my intention. If SHELL is selected I assume it is used, hence this part of code is needed.
I apologize if I was not clear.

**Comment by andyross** [comment]
Sorry, why can't the code (presumably logging/output code in the shell?) be modified so that it does the same thing in all cases (e.g. "format the output string using sprintf") and not have to be in an situation where sometimes it uses printk (which is non-standard and lacks support for some things that apparently shell wants) and sometimes it uses libc printf (which should be reasonably ISO C compliant).

Basically the bug here seems to be that printk() was used at all (i.e. for purposes other than kernel logging and debugging), not that it lacks support for obscure features.
